c/client.c.go

```diff
@@ -45,160 +35,84 @@ type transferContext struct {
 }
 
 type clientWithContext struct {
-	appContext Application
+	appContext PendingTransfer
 	client     *wormhole.Client
 }
 
 var (
 	ErrClientNotFound = fmt.Errorf("%s", "wormhole client not found")
 
-	clientsMap       map[int32]clientWithContext = map[int32]clientWithContext{}
-	pendingTransfers map[int]transferContext     = map[int]transferContext{}
-	transfersLock    sync.Mutex
+	pendingTransfers map[unsafe.Pointer]transferContext = map[unsafe.Pointer]transferContext{}
 )
 
-func removePendingTransfer(transferId int) {
-	transfersLock.Lock()
-	defer transfersLock.Unlock()
-
-	close(pendingTransfers[transferId].Commands)
-	delete(pendingTransfers, transferId)
-}
-
-func addPendingTransfer(cancelFunc context.CancelFunc) int {
-	transfersLock.Lock()
-	defer transfersLock.Unlock()
-
-	transferId := len(pendingTransfers)
-	pendingTransfers[transferId] = transferContext{
+func addPendingTransfer(transferRef unsafe.Pointer, cancelFunc context.CancelFunc) {
+	pendingTransfers[transferRef] = transferContext{
 		Commands:   make(chan Command),
 		CancelFunc: cancelFunc,
 	}
-
-	return transferId
-}
-
-//export NewClient
-func NewClient(appId *C.char, rendezvousUrl *C.char, transitRelayUrl *C.char, passPhraseComponentLength C.int) int32 {
-	client := &wormhole.Client{
-		AppID:                     DEFAULT_APP_ID,
-		RendezvousURL:             DEFAULT_RENDEZVOUS_URL,
-		TransitRelayURL:           DEFAULT_TRANSIT_RELAY_URL,
-		PassPhraseComponentLength: DEFAULT_PASSPHRASE_COMPONENT_LENGTH,
-	}
-
-	if appId != nil {
-		client.AppID = C.GoString(appId)
-	}
-
-	if rendezvousUrl != nil {
-		client.RendezvousURL = C.GoString(rendezvousUrl)
-	}
-
-	if transitRelayUrl != nil {
-		client.TransitRelayURL = C.GoString(transitRelayUrl)
-	}
-
-	if passPhraseComponentLength == 0 {
-		client.PassPhraseComponentLength = int(passPhraseComponentLength)
-	}
-
-	clientId := int32(uintptr(unsafe.Pointer(client)))
-	clientsMap[clientId] = clientWithContext{
-
-		client: client,
-	}
-
-	return clientId
 }
 
 //export Finalize
-func Finalize(clientId int32) int32 {
-	client, ok := clientsMap[clientId]
-
-	client.appContext.Log("Finalizing client with id %d", clientId)
-
+func Finalize(transfer *C.wrapped_context_t) {
+	transfer.Log("Finalizing transfer: %p", transfer)
+	transferContext, ok := pendingTransfers[unsafe.Pointer(transfer)]
 	if !ok {
-		return int32(codes.ERR_NO_CLIENT)
+		panic("Finalizing an invalid transfer")
```

Comment:
I wonder whether `panic()` is a good idea for a library. How does it translate across the border to C? Are there ways to detect such a panic and stop the program than just crashing?

Reply:
I think this is similar in severity to calling `free()` on an invalid pointer. So it would be a `SIGABRT` for the process. I believe only a buggy application would trigger that case, but I'm not sure if we should instead fail softly with a return value. :thinking: